File: solutions/Medium/Lowest_Common_Ancestor_of_a_Binary_Tree_0236.py
from typing import Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root: Optional[TreeNode], p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
        """
        Function: lowestCommonAncestor
        Description: Finds the lowest common ancestor (LCA) of two given nodes in a binary tree.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.
        - p (TreeNode): The first target node.
        - q (TreeNode): The second target node.

        Returns:
        - TreeNode: The lowest common ancestor of p and q.
        """
        if not root or root == p or root == q:
            return root

        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)

        if left and right:
            return root  # If p and q are found in different subtrees, root is LCA

        return left if left else right  # Return the non-null subtree


# Helper functions for testing
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = {val: TreeNode(val) for val in lst if val is not None}
    root = nodes[lst[0]]

    for i, val in enumerate(lst):
        if val is not None:
            node = nodes[val]
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst) and lst[left_idx] is not None:
                node.left = nodes[lst[left_idx]]
            if right_idx < len(lst) and lst[right_idx] is not None:
                node.right = nodes[lst[right_idx]]

    return root, nodes


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: LCA of 5 and 1 is 3
    tree, nodes = list_to_binary_tree([3, 5, 1, 6, 2, 0, 8, None, None, 7, 4])
    print(solution.lowestCommonAncestor(tree, nodes[5], nodes[1]).val)
    # Expected output: 3

    # Test case 2: LCA of 5 and 4 is 5
    tree, nodes = list_to_binary_tree([3, 5, 1, 6, 2, 0, 8, None, None, 7, 4])
    print(solution.lowestCommonAncestor(tree, nodes[5], nodes[4]).val)
    # Expected output: 5

    # Test case 3: LCA of 1 and 2 is 1
    tree, nodes = list_to_binary_tree([1, 2])
    print(solution.lowestCommonAncestor(tree, nodes[1], nodes[2]).val)
    # Expected output: 1

    # Test case 4: LCA of 1 and 1 is 1
    tree, nodes = list_to_binary_tree([1])
    print(solution.lowestCommonAncestor(tree, nodes[1], nodes[1]).val)
    # Expected output: 1

    # Test case 5: LCA of 1 and 2 is 1
    tree, nodes = list_to_binary_tree([1, 2, 3])
    print(solution.lowestCommonAncestor(tree, nodes[1], nodes[2]).val)
    # Expected output: 1

    # Test case 6: LCA of 1 and 3 is 1
    tree, nodes = list_to_binary_tree([1, 2, 3])
    print(solution.lowestCommonAncestor(tree, nodes[1], nodes[3]).val)
    # Expected output: 1

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(n), where n is the number of nodes in the binary tree. In the worst case, we might visit all the nodes of the binary tree.
# Space Complexity: O(n). The maximum amount of space utilized by the recursion stack would be n since the height of a skewed binary tree could be n.

File: solutions/Medium/Lowest_Common_Ancestor_of_a_Binary_Tree_0236.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Function: lowestCommonAncestor
 * Description: This function finds the lowest common ancestor (LCA) of two given nodes
 * in a binary tree. The LCA is the lowest node in the tree that has both `p` and `q`
 * as descendants (including the node itself).
 *
 * @param {TreeNode} root - The root of the binary tree.
 * @param {TreeNode} p - The first target node.
 * @param {TreeNode} q - The second target node.
 * @returns {TreeNode} - The lowest common ancestor of `p` and `q`.
 */
const lowestCommonAncestor = function(root, p, q) {
    // Base case: if root is null, or root matches either p or q, return root
    if (!root || root === p || root === q) {
        return root;
    }

    // Recursively search for LCA in left and right subtrees
    let left = lowestCommonAncestor(root.left, p, q);
    let right = lowestCommonAncestor(root.right, p, q);

    // If both left and right are non-null, root is the LCA
    if (left && right) {
        return root;
    }

    // If only one side is non-null, return that side (either left or right)
    return left || right;
};

// Example Test Cases
if (require.main === module) {
    function TreeNode(val) {
        this.val = val;
        this.left = this.right = null;
    }

    // Construct Example Tree: [3,5,1,6,2,0,8,null,null,7,4]
    let root = new TreeNode(3);
    root.left = new TreeNode(5);
    root.right = new TreeNode(1);
    root.left.left = new TreeNode(6);
    root.left.right = new TreeNode(2);
    root.right.left = new TreeNode(0);
    root.right.right = new TreeNode(8);
    root.left.right.left = new TreeNode(7);
    root.left.right.right = new TreeNode(4);

    console.log(lowestCommonAncestor(root, root.left, root.right).val);
    // Expected output: 3 (LCA of 5 and 1)

    console.log(lowestCommonAncestor(root, root.left, root.left.right.right).val);
    // Expected output: 5 (LCA of 5 and 4)

    let root2 = new TreeNode(1);
    root2.left = new TreeNode(2);
    console.log(lowestCommonAncestor(root2, root2, root2.left).val);
    // Expected output: 1 (LCA of 1 and 2)
}

/*

Complexity Analysis:
- Time Complexity: O(n) - We visit each node at most once.
- Space Complexity: O(n) - In the worst case (skewed tree), recursion depth reaches O(n).

*/

export default lowestCommonAncestor;
